### PR TITLE
Require default role to be set for using connection impersonation

### LINKED ISSUE #33423: Change in Snowflake Driver adds a default role ACCOUNTADMIN instead of using user default
### Describe the bug

- Connecting to Snowflake without specifying role
- When we added connection impersonation, we added a default role of ACCOUNTADMIN that most Snowflake users don't or should not have in major deployments instead of allowing Snowflake user's default role to be used like many of our customers expect (and is the standard for other connections).

### To Reproduce

1. Use credentials on Snowflake that don't have ACCOUNTADMIN role assigned
2. Admin > Databases > Add > Fill this out, but leave Role blank
3. Try to connect to database and see error `(Requested role 'ACCOUNTADMIN' is not assigned to the executing user. Specify another role to activate.)`

### Expected behavior

Should use the default assigned to the user if left blank.

### Logs

image

### Information about your Metabase installation

```JSON
- Metabase: v1.47.0
```

### Severity

P1: Regression preventing connecting using user default role

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by noahmoss**
Updates the connection impersonation configuration process for Snowflake databases to require `role` to be filled out in the database details page. This allows Metabase to determine what role should be used to reset connections to for non-impersonated users, since there's no easy way to obtain that from Snowflake (unlike with Postgres). Discussion is in [this thread](https://metaboat.slack.com/archives/C056QEW4KNF/p1692795342224929).

This also includes a small frontend change & a new unit test to show a different warning in the impersonation modal if the `role` field is not filled out for a Snowflake DB.

Resolves https://github.com/metabase/metabase/issues/33423

### CODE REVIEW

File: enterprise/backend/test/metabase_enterprise/advanced_permissions/driver/impersonation_test.clj [hunk 1]
```diff
@@ -95,15 +96,37 @@
...
       (advanced-perms.api.tu/with-impersonations {:impersonations [{:db-id (mt/id) :attribute "impersonation_attr"}]
                                                   :attributes     {"impersonation_attr" "limited_role"}}
-        ;; User with connection impersonation should not be able to query a table they don't have access to
-        ;; (`limited_role` in CI Snowflake has no data access)
-        (is (thrown-with-msg? clojure.lang.ExceptionInfo
-                              #"SQL compilation error:\nDatabase.*does not exist or not authorized"
-                              (mt/run-mbql-query venues
-                                                 {:aggregation [[:count]]})))
-        ;; Non-impersonated user should stil be able to query the table
+        ;; Test database initially has no default role set. All queries should fail, even for non-impersonated users,
```
Comment (adam-james-v) [inline enterprise/backend/test/metabase_enterprise/advanced_permissions/driver/impersonation_test.clj:99]: Is this true for every db driver? I thought part of this was that some drivers _do_ have default roles that work without configuring them. So, would this part of the test fail for one of those dbs now?

**Review by adam-james-v — APPROVED** [review]
Looks good to me. Just left one question about whether one assertion in the test will work with other drivers or not. Please ignore if I just misunderstood!

File: enterprise/backend/test/metabase_enterprise/advanced_permissions/driver/impersonation_test.clj (hunk 1 above)
Comment (noahmoss) [reply enterprise/backend/test/metabase_enterprise/advanced_permissions/driver/impersonation_test.clj:99]: This test is only for Snowflake! Postgres (the only other driver that can use conn impersonation at the moment) has a separate test above

Comment (adam-james-v) [reply enterprise/backend/test/metabase_enterprise/advanced_permissions/driver/impersonation_test.clj:99]: Ah, oops. Sorry for the noise. Just missed it my bad.

File: modules/drivers/snowflake/resources/metabase-plugin.yaml [hunk 2]
```diff
@@ -33,7 +33,7 @@ driver:
       type: schema-filters
       display-name: Schemas
     - name: role
-      display-name: Role (optional)
+      display-name: Role (required for connection impersonation)
```
Comment (iethree) [inline modules/drivers/snowflake/resources/metabase-plugin.yaml:36]: 👍

**Review by iethree — CHANGES_REQUESTED** [review]
Frontend code looks good. I have a misgiving about the way we check for whether role is required though.

**Review by iethree — APPROVED** [review]
LGTM 🥳

**Pull request merged**